### PR TITLE
Create iam.py, consistent naming for IAM functions

### LINKED ISSUE #680: Naming pattern for IAM functions
**Is your feature request related to a problem? Please describe.**
It may be easier for users to find and select IAM functions if the function names start with the prefix `iam`, e.g., `iam_ashrae` rather than the current pattern placing `iam` as a suffix `ashraeiam`

**Describe the solution you'd like**
Rename `physicaliam`, `ashraeiam`, etc. and deprecate old names.

### PR DESCRIPTION
**Pull request by cwhanse**
- [x] Closes  #680. Related to #436 
 - [x] I am familiar with the [contributing guidelines](http://pvlib-python.readthedocs.io/en/latest/contributing.html).
 - [x] Fully tested. Added and/or modified tests to ensure correct behavior for all reasonable inputs. Tests (usually) must pass on the TravisCI and Appveyor testing services.
 - [x] Updates entries to `docs/sphinx/source/api.rst` for API changes.
 - [x] Adds description and name entries in the appropriate `docs/sphinx/source/whatsnew` file for all changes.
 - [x] Code quality and style is sufficient. Passes LGTM and SticklerCI checks.
 - [x] New code is fully documented. Includes sphinx/numpydoc compliant docstrings and comments in the code where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.

Brief description of the problem and proposed solution (if not already fully described in the issue linked to above):

- Functions that calculate incident angle modifier IAM are moved from `pvsystem.py` to the new `iam.py`.
- Function names are shorted to only the model, e.g., `iam.ashrae` instead of `pvsystem.ashraeiam`.
- Method names for `PVSystem are prepended by `iam_`, e.g., `PVSystem.iam_ashrae` instead of `PVSystem.ashrae_iam`.
- The SAPM IAM function is renamed from `pvsystem.sapm_aoi_loss` to `iam.sapm`.
- Edits to function docstrings.

### CODE REVIEW

**Comment by wholmgren** [comment]
> Method names for PVSystem are prepended by iam_, e.g., PVSystem.iam_ashraeinstead ofPVSystem.ashrae_iam`.

What about a single `PVSystem.iam` wrapper method? Might present a cleaner API.

**Comment by cwhanse** [comment]
> Method names for PVSystem are prepended by iam_, e.g., PVSystem.iam_ashraeinstead ofPVSystem.ashrae_iam`.
> 
> What about a single PVSystem.iam wrapper method? Might present a cleaner API.

Do you think we should start using .setter properties the same way we do in ModelChain?

**Comment by cwhanse** [comment]
Ready for review. Notes on the scope of this PR and the test failure:

- `iam.interp` is not connected to `PVSystem` or `ModelChain`. I didn't think of a convenient way to pass the `interp` unique arguments through `PVSystem.get_iam`.
- `get_psm3` is failing due to a change in data on the NSRDB. `get_psm3` is fetching pressure in units of decibars, the recorded test file (and NSRDB documentation) says mbar. I've sent a note to NREL asking about the change, which I think may be in error, since TMY data fetched through the web interface is in mbar. Otherwise it appears that `get_psm3` works as intended.

File: pvlib/iam.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Wed Oct  2 09:04:01 2019
+
+@author: cwhanse
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply pvlib/iam.py]: Leftover auto code from my IDE, I'll delete and replace with a module-level docstring.  That's a good idea.

File: pvlib/iam.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+"""
+import numpy as np
+import pandas as pd
+from pvlib.tools import cosd, sind, tand, asind
+
+
+# a dict of required parameter names for each IAM model
+# keys are the function names for the IAM models
+IAM_MODEL_PARAMS = {
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply pvlib/iam.py]: I'm also uneasy about continued use of constants following this pattern. I think it's valuable to have lists of model parameter names somewhere, perhaps not in this form.  A data model encompassing a system's physical and model parameters would help here.

File: pvlib/test/test_pvsystem.py [hunk 3]
```diff
@@ -82,162 +79,40 @@ def test_systemdef_dict():
...
+def test_PVSystem_get_iam(mocker, iam_model, model_params):
+    m = mocker.spy(_iam, iam_model)
+    module_parameters = pd.Series(model_params)
     system = pvsystem.PVSystem(module_parameters=module_parameters)
-    mocker.spy(pvsystem, 'physicaliam')
     thetas = 1
-    iam = system.physicaliam(thetas)
-    pvsystem.physicaliam.assert_called_once_with(thetas, **module_parameters)
+    iam = system.get_iam(thetas, iam_model=iam_model)
+    assert m.call_count == 1
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply pvlib/test/test_pvsystem.py]: Is the mistake obvious in this? I can't see it.  From the log it appears that the mock `m` isn't being set on the desired function:

```
@pytest.mark.parametrize('iam_model,model_params', [
    ('ashrae', {'b': 0.05}),
    ('physical', {'K': 4, 'L': 0.002, 'n': 1.526}),
    ('martin_ruiz', {'a_r': 0.16}),
])
def test_PVSystem_get_iam(mocker, iam_model, model_params):
    m = mocker.spy(_iam, iam_model)
```

File: pvlib/test/test_pvsystem.py [hunk 4]
```diff
@@ -82,162 +79,39 @@ def test_systemdef_dict():
...
+    ('martin_ruiz', {'a_r': 0.16}),
+])
+def test_PVSystem_get_iam(mocker, iam_model, model_params):
+    m = mocker.spy(_iam, iam_model)
+    system = pvsystem.PVSystem(module_parameters=model_params)
     thetas = 1
-    iam = system.physicaliam(thetas)
-    pvsystem.physicaliam.assert_called_once_with(thetas, **module_parameters)
+    iam = system.get_iam(thetas, iam_model=iam_model)
+    assert m.assert_called_once_with(thetas, **model_params)
```
Comment (wholmgren) [inline pvlib/test/test_pvsystem.py]: The problem is the extra `assert` at the beginning of the line. Python first calls `m.assert_called_once_with(thetas, **model_params)`, which returns `None` if it does not error. Then the line looks like `assert None`. which always fails.

**Comment by wholmgren** [comment]
Might want to mark `test_get_psm3` with pytest's `xfail` decorator.

**Comment by cwhanse** [comment]
> Might want to mark test_get_psm3 with pytest's xfail decorator.

I spoke with NREL a few days back, they confirmed there's an error on their end with the pressure units, and they're working on it.

File: pvlib/iam.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wholmgren) [reply pvlib/iam.py]: Created #789 to track this. I'm ok leaving it alone for this PR.

**Comment by wholmgren** [comment]
[this line](https://codecov.io/gh/pvlib/pvlib-python/compare/a210432595839f3efbbf669d5d6b1dbbd4b1ff9f...23f2677ace8db33f479b7438fb28422818840072/diff#D2-476) is not covered. I think it will be if you move the `assert` out from the `with` block. That block ends when the exception is raised, then the function continues running past the block (currently running nothing, but should be running the assert).

**Pull request merged**